### PR TITLE
feat: Group folder support

### PR DESCRIPTION
**Pull request by WinterOfBit**
## Support group folder feature.
frontend pr: https://github.com/cloudreve/frontend/pull/168
solved: #1305 #1673 

### Important: key db model modified.
used some little tricks to impl this function and minimize impact meanwhile

models/file.go:
```go
type File struct {
    ...
    UserID int // previous uint
    ...
}
```

models/folder.go:
```go
type Folder struct {
    ...
    OwnerID int // previous uint
    ...
}
```

After this pr, type of `userId` , `ownerId` or any `id` that indicates user owner is `int`. Positive `id` means a `user` owner. Negative `id` means its owner is a `group`

Will upload demo video later. Hope this pr works for those who needed group folder feature before v4 came out : )

### CODE REVIEW

**Comment by WinterOfBit** [comment]
Enable & disable group folder:

https://github.com/cloudreve/Cloudreve/assets/68735689/4190dd50-95df-4ac7-bb17-62ffbbccc058

Upload & download files into group folder:

https://github.com/cloudreve/Cloudreve/assets/68735689/fba99d4a-cf34-406d-b7d6-b42824451f6f

plz review

File: models/file.go [hunk 1]
```diff
@@ -63,8 +63,12 @@ func (file *File) Create() error {
 		return err
 	}
 
+	if file.UserID < 0 {
+		return tx.Commit().Error
+	}
```
Comment (HFO4) [inline models/file.go:68]: Unlimited storage for group folders?

File: models/group.go [hunk 2]
```diff
@@ -82,3 +100,47 @@ func (group *Group) SerializePolicyList() (err error) {
...
+
+func (group *Group) getGroupFolder() (*Folder, bool) {
+	var folder Folder
+	result := DB.Where("owner_id = ?", -int(group.ID)).First(&folder)
+	return &folder, !result.RecordNotFound()
+}
+
+// createGroupFolder 创建用户组文件夹，如果已存在则更新
+func (group *Group) createGroupFolder(name string) error {
+	tx := DB.Begin()
```
Comment (HFO4) [inline models/group.go:112]: TX inside of ongoing TX, this will be a deadlock under SQLite.

File: models/folder.go [hunk 3]
```diff
@@ -15,7 +15,7 @@ type Folder struct {
 	gorm.Model
 	Name     string `gorm:"unique_index:idx_only_one_name"`
 	ParentID *uint  `gorm:"index:parent_id;unique_index:idx_only_one_name"`
-	OwnerID  uint   `gorm:"index:owner_id"`
+	OwnerID  int    `gorm:"index:owner_id"`
```
Comment (HFO4) [inline models/folder.go:18]: Existing operations on folders will break on group folders. E.g. query folder props, renaming folder.

Comment (HFO4) [reply models/folder.go:18]: Create share link on group folder does not work.

Comment (HFO4) [reply models/folder.go:18]: Move/Copy inside of a group folder, path selector does not work.
"Uploader" is shown as blank in the admin dashboard.

File: pkg/filesystem/manage.go [hunk 4]
```diff
@@ -288,6 +287,12 @@ func (fs *FileSystem) List(ctx context.Context, dirPath string, pathProcessor fu
 	// 获取子文件
 	childFiles, _ = folder.GetChildFiles()
 
+	if dirPath == "/" {
+		if fs.User.Group.GroupFolder != nil {
+			childFolders = append(childFolders, *fs.User.Group.GroupFolder)
```
Comment (HFO4) [inline pkg/filesystem/manage.go:292]: What if users already have duplicated folder or file under root?

File: models/group.go [hunk 5]
```diff
@@ -82,3 +100,47 @@ func (group *Group) SerializePolicyList() (err error) {
...
+		}
+	}
+
+	group.GroupFolder = folder
+	return tx.Commit().Error
+}
+
+// deleteGroupFolder 删除用户组文件夹
+func (group *Group) deleteGroupFolder() {
+	tx := DB.Begin()
```
Comment (HFO4) [inline models/group.go:135]: What about existing file/folders under group folder?

**Review by HFO4 — COMMENTED** [review]
Thanks for the PR! Really appreciate your effort to make this complex feature being reality. Sadly, I cannot merge this change into the main branch. But feel free to create a new forked version with this feature!
The basic idea of using one DB field for two models (User/Group) is innovative, but the tradeoff is breaking too many existing features. It is not compatible with the shared folder design in V4 (which is already implemented), which means it will be really challenging to upgrade from V3 to V4 in the future.

**Comment by WinterOfBit** [comment]
> Thanks for the PR! Really appreciate your effort to make this complex feature being reality. Sadly, I cannot merge this change into the main branch. But feel free to create a new forked version with this feature!
> The basic idea of using one DB field for two models (User/Group) is innovative, but the tradeoff is breaking too many existing features. It is not compatible with the shared folder design in V4 (which is already implemented), which means it will be really challenging to upgrade from V3 to V4 in the future.
> 

Thanks for ur review! You're right that my impl is quite tricky and incomplete. And considering the difficulity brought when migarating to v4, It's true wiser to make this a forked version instead of merging.

**Comment by WinterOfBit** [comment]
By the way, will v4 be open-source? Can community join development or try alpha version?

**Comment by HFO4** [comment]
> By the way, will v4 be open-source? Can community join development or try alpha version?

It will surely be open-source. But shared folder will be a limited feature in Pro version. Current V4 development is in a private Pro version repo. Community contribution is welcomed after the first V4 alpha version is shipped.

File: models/file.go [hunk 6]
```diff
@@ -20,7 +20,7 @@ type File struct {
 	gorm.Model
 	Name            string `gorm:"unique_index:idx_only_one"`
 	SourceName      string `gorm:"type:text"`
-	UserID          uint   `gorm:"index:user_id;unique_index:idx_only_one"`
+	UserID          int    `gorm:"index:user_id;unique_index:idx_only_one"`
```
Comment (zzti) [inline models/file.go:23]: 为什么这里要把uint改为int呢?

**Pull request closed without merging**